### ISSUE TITLE
Inconsistent feeling handling of Any in binder

### ISSUE THREAD

**Issue by hauntsaninja** [python/mypy#20870]
```
from typing import Any, Final

CONSTANT: Final = 2015

def good(result: dict[str, Any]) -> None:
    code = 'asdf'
    if isinstance(result, dict):
        code = result.get('code', code)
    reveal_type(code)  # N: Revealed type is "Any"
    if code == CONSTANT:
        reveal_type(code)  # N: Revealed type is "Any"

def bad(result: Any) -> None:
    code = 'asdf'
    if isinstance(result, dict):
        code = result.get('code', code)
    reveal_type(code)  # N: Revealed type is "builtins.str"
    if code == CONSTANT:
        reveal_type(code)  # N: Revealed type is "builtins.str"
```

Based on real world code for rotki

Not sure what if anything should be done here. But it feels a little inconsistent / violating of gradual typing

Found this via https://github.com/python/mypy/pull/20660 which makes `--warn-unreachable` flag raise a diagnostic on the second comparison

**Comment by ilevkivskyi** [comment]
I vaguely remember discussion with @cdce8p about this, and IIRC he said that your `bad()` is actually good (and vice versa). The current behavior w.r.t. `Any` in binder is a result of many compromises. At some point I tried to make it proper and principled (which is actually not hard by itself), but I lost, because people were not happy with getting more possible false negatives.

**Comment by ilevkivskyi** [comment]
Btw this works (much) better with `--allow-redefinition-new`. For last two reveals I am getting
```
test.py:19: note: Revealed type is "builtins.str | Any"
test.py:21: note: Revealed type is "Any | Literal[2015]?"
```
